FormSchemaEditor component for editing form layout

legitmaxwu opened the pull request:
* Added NoSSR wrapper for components that shouldn't be server-rendered
* Created Form-related Typescript types

nova-admin commented:
#### 📝Changed routes:


| [**/test/testformeditor**](https://risc-1s97mkzu6-novaforgood.vercel.app/test/testformeditor) |
|:-:|
|<a href="https://risc-1s97mkzu6-novaforgood.vercel.app/test/testformeditor"><img src="https://images.weserv.nl?url=https://sjc.microlink.io/EkQec-f2CJeyulUp61WtBa4yfMwK3BrDvSnmsiD-v9F9Qaa4FoFl-5oqNtWZuc6i8ZWywdjoVQ9pbMTt0G1OFQ.png&w=600" alt="Screenshot of /test/testformeditor" width="300"></a><br /><sup><a href="https://images.weserv.nl?url=https://sjc.microlink.io/af8NWL372HTuqQH8bgO2yfIRht2jaFKahbZ66nhLv8v3ZfkIs2dTrHPlwG3WZ0i9TBk9arTe_QGxB95AxSNn-w.png">(view full size)</a> |







Commit 8f52dd812508c74a3dd3c4232b05b58a564f9880 (https://risc-1s97mkzu6-novaforgood.vercel.app).

samanthachai commented:
**Possible Changes**
- To prevent shifting, I think the extra border/blur should be removed. Darkening the outline alone should be fine. 
- Click to edit is not clear, would be nice to have an edit indicator. Maybe hover to edit or an actual edit button. 
- Edge case: If the question or _helper text_ is too long, it goes outside of the border. 
- "Paragraph" vs "Single Line" is vague + unclear. Maybe add "Response" after. 
- When editing the last question, there could be more spacing between the end of the question and the top of the button. 

**Good Things**
- I like the drag/move icon. 
- Hover color is not distracting.

CarterBian commented:
- very clean liked the edit form design that popped up after click
- was a little confused at first about editing so some sort of edit indicator without hovering over the question like sam said would be nice

duncandonuts6 commented:
- Pretty clean for the most part.
- Add Question is pretty close to the bottom of the page once you make the form long enough. Maybe bump up a bit?
- Maybe, instead of Description, the user could put an example answer / answer directions in the answer spot so that the form filler outer would have some inspo / know what format to put responses in. I was personally struggling with what else to put in the Description field.

serene-sup commented:
Feedback:

1. weird that everything shifts slightly when u hover over an input field (assuming its bc the box stroke thickens w/ hover); change suggestion: just change border color w/out width

2. is there a reason why you can type in the input field in preview view? feels strange bc can’t do anything w/ the text typed. also im inclined to click the input field 1st instead of other parts in that section/“question” (even tho entire section color changes w/ hover) so makes me think I type the question in the box

2.5 input field filled in edit mode?? 1) what r u tryna communicate to the user by having the input field shown in edit view (nothing wrong w/ it being there, just curious bc im not sure that is getting across rn) 2) is there a reason why it’s filled in w/ grey? 1st impression when i see if after clicking to edit is ?wot is dis box?

3. “helper text” is confusing. even calling it “description” feels confusing but prolly a bit better. makes sense on what that field is after i play around w/ it a bit but the label doesn’t feel like it describes the functionality of that part too well 

4. minor thing kinda nitpicky but a lot of white space between question and text field below it

5. another nitpicky minor ui thing might be a stylistic preference but i think description text should have more contrast from question text maybe use smaller font size 

Other thoughts:
- i like the direction this is going in! its good to be able to see a preview of how mentors/mentees would see it on the other side + the editor view structure is a good idea
- is this the final ui design cuz i have some other stylistic comments but if u r just looking at functionality rn i can withhold
- thanks for working on disssss!!!

legitmaxwu commented:
Thanks for the detailed feedback everyone!

@samanthachai @CarterBian 
- I think I addressed all your points in the newest version

@duncandonuts6 
Leaving a todo:
- [ ] Question: should we replace setting description with setting answer placeholder

@serene-sup 
- For 1-3, I pushed fixes so check the newest version!
- Focusing on UX first, but if you have comments don't withhold! I can just record it on this PR thread.

Leaving todos:
- [ ] Too much white space between question text field and description text field
- [ ] Smaller font size for description text

innopoop commented on src/components/FormSchemaEditor/FormQuestionSchemaEditor.tsx:
```diff
@@ -0,0 +1,237 @@
+                <div className="w-2"></div>
+
+                <div className="w-1"></div>
+                <div className="rounded hover:bg-tertiary p-1.5 cursor-pointer">
```
Pressing on the outer part of the delete icon (hover background) does not initiate onClick function which is confusing

legitmaxwu replied:
Nice catch--thanks!

innopoop commented on src/components/FormSchemaEditor/Select.tsx:
```diff
@@ -0,0 +1,100 @@
+import { Fragment } from "react";
+
+interface UpDownArrowProps {}
+const UpDownArrow: React.FC<UpDownArrowProps> = () => {
```
On a previous commit of mine, @kevinl120 commented about the downsides of React.FC- I don't know if it also applies here. [reference](https://github.com/facebook/create-react-app/pull/8177) and [reference2](https://www.harrymt.com/blog/2020/05/20/react-typescript-react-fc.html)

innopoop commented on src/components/FormSchemaEditor/Select.tsx:
```diff
@@ -0,0 +1,100 @@
+    <div className="w-auto">
+      <Listbox
+        value={value}
+        onChange={(selectedValue) => {
```
Implicitly has type any! Along with lines 74 & 81

innopoop commented on src/components/FormSchemaEditor/index.tsx:
```diff
@@ -0,0 +1,151 @@
+
+// Types
+
+export type RecursivePartial<T> = {
```
Does this end up being used anywhere besides in utils.tsx?

legitmaxwu replied:
Probably, but we can move the definition somewhere better in the future I think

innopoop reviewed (changes requested):
Main thing is that clicking on the delete hover area (not the icon) doesn't delete the question

innopoop reviewed (approved):
LGTM LGTM LGTM

innopoop reviewed (approved):
This is the one I'm supposed to LGTM mb
